chia/data_layer/data_store.py

```diff
@@ -0,0 +1,138 @@
+
+# from chia.types.full_block import FullBlock
+# from chia.types.weight_proof import SubEpochChallengeSegment, SubEpochSegments
+from chia.util.db_wrapper import DBWrapper
+
+# from chia.util.ints import uint32
+# from chia.util.lru_cache import LRUCache
+
+log = logging.getLogger(__name__)
+
+
+class OperationType(IntEnum):
+    INSERT = 0
+    DELETE = 1
+
+
+@dataclass(frozen=True)
+class Action:
+    type: OperationType
+    row_index: int
+    row: CLVMObject.CLVMObject
+
+
+@dataclass(frozen=True)
+class Commit:
+    # actions: OrderedDict[bytes32, CLVMObject.CLVMObject]
+    actions: Tuple[Action, ...]
+    changelist_hash: bytes32
+    # TODO: bytes32 may be totally wrong here for the merkle root hash of the overall
+    #       data state.
+    root_hash: bytes32
+
+    @classmethod
+    def build(cls, actions: Iterable[Action], root_hash: bytes32) -> "Commit":
+        # TODO: calculate the hash
+        changelist_hash = bytes32()
+
+        actions = tuple(actions)
+
+        return cls(actions=actions, changelist_hash=changelist_hash, root_hash=root_hash)
+
+
+class DataStore:
+    db: aiosqlite.Connection
+    # block_cache: LRUCache
+    db_wrapper: DBWrapper
+    # ses_challenge_cache: LRUCache
+
+    @classmethod
+    async def create(cls, db_wrapper: DBWrapper):
+        self = cls()
+
+        # All full blocks which have been added to the blockchain. Header_hash -> block
+        self.db_wrapper = db_wrapper
+        self.db = db_wrapper.db
+
+        # TODO: what pragmas do we want?
+        # await self.db.execute("pragma journal_mode=wal")
+        # await self.db.execute("pragma synchronous=2")
+
+        # TODO: make this handle multiple data layer tables
+        # TODO: do we need to handle multiple equal rows
+
+        # Just a raw collection of all ChiaLisp lists that are used
+        await self.db.execute("CREATE TABLE IF NOT EXISTS raw_rows(row_hash TEXT PRIMARY KEY, clvm_object BLOB)")
+        # The present properly ordered collection of rows.
+        await self.db.execute("CREATE TABLE IF NOT EXISTS data_rows(row_hash TEXT PRIMARY KEY)")
+        # TODO: needs a key
+        # TODO: As operations are reverted do they get deleted?  Or perhaps we track
+        #       a reference into the table and only remove when a non-matching forward
+        #       step is taken?  Or reverts are just further actions?
+        await self.db.execute(
+            "CREATE TABLE IF NOT EXISTS actions(" "data_row_index INTEGER, row_hash TEXT, operation INTEGER" ")"
+        )
+        # TODO: Could also be structured such that the action table has a reference from
+        #       each action to the commit it is part of.
+        await self.db.execute("CREATE TABLE IF NOT EXISTS commits(" "changelist_hash TEXT, actions_index INTEGER" ")")
+
+        await self.db.commit()
+
+        return self
+
+    # TODO: Add some handling for multiple tables.  Could be another layer of class
```

Comment:
we can make each row can contain a "Table" column containing the id of the table it belongs to.
that way we can manage adding an removing tables easy

Reply:
Agreed, I meant multiple data layer tables, not multiple sql tables.